### PR TITLE
Separate Configuration for Multiple Workers.

### PR DESCRIPTION
**Pull request by dartjason**
add multi track training feature

### CODE REVIEW

**Comment by larsll** [comment]
@dartjason - I was thinking; instead of having a lot of additional parameters, could we not simply do a comma-separated value in WORLD_NAME?

**Comment by dartjason** [comment]
Yeah sure we can do that.  I do think it’s a bit harder to read that way though.

**Comment by larsll** [comment]
I guess my idea would be to do more of a 'multi-config'; I essentially with all relevant parameters being possible to have in one or NUM_WORKERS versions? Multiple tracks, multiple modes (e.g. 2x Time Trial + 2x H2B). That would need some more thinking as to what the best approach is.

**Comment by larsll** [comment]
Been thinking a bit more; maybe a worker-x.env would be the thing; there we can put in all potential worker related options which can then be overwritten. The prepare-config.py will get more complicated but it would be worth it.

**Comment by dartjason** [comment]
I agree this would provide a lot of flexibility to change other parameters for each worker.  I guess question is do we envision changing others besides the track?

**Comment by larsll** [comment]
I want to change every parameter; for head-to-bot I would like to have one doing TT, one doing OA with the boxes looking like cars, and two doing H2B with different parameters

**Comment by dartjason** [comment]
Makes sense.  I can try to carve out some time this week to work on it.

**Comment by dartjason** [comment]
Updated to allow for multiple training-param configs.  I think this is good to go based on the testing I conducted, but please have another look and maybe do some testing.

File: scripts/training/prepare-config.py [hunk 1]
```diff
@@ -86,3 +86,91 @@
     yaml.dump(config, yaml_file, default_flow_style=False, default_style='\'', explicit_start=True)
 
 s3_client.upload_file(Bucket=s3_bucket, Key=yaml_key, Filename=local_yaml_path)
+
+
+# Training with different configurations on each worker (aka Multi Config training)
+config['MULTI_CONFIG'] = os.environ.get('DR_MULTI_CONFIG', 'False')
```
Comment (larsll) [inline scripts/training/prepare-config.py]: Let's call the variable DR_TRAIN_MULTI_CONFIG as it only applies to training.

File: scripts/training/prepare-config.py [hunk 2]
```diff
@@ -86,3 +86,91 @@
...
+        if i == 1:
+            # copy training_params to training_params_1
+            s3_yaml_name_list = s3_yaml_name.split('.')
+            s3_yaml_name_temp = s3_yaml_name_list[0] + "_%d.yaml" % i
+
+            #upload additional training params files
+            yaml_key = os.path.normpath(os.path.join(s3_prefix, s3_yaml_name_temp))
+            s3_client.upload_file(Bucket=s3_bucket, Key=yaml_key, Filename=local_yaml_path)
+
+            file1 = open("multiconfig.txt", "w")
```
Comment (larsll) [inline scripts/training/prepare-config.py]: File should be stored in the DR_DIR/tmp directory -- preferrably prefixed with the DR_RUN_ID not to break multiple runs in parallel.

File: scripts/training/start.sh [hunk 3]
```diff
@@ -69,11 +69,40 @@ if [ "$DR_WORKERS" -gt 1 ]; then
...
+    export ROBOMAKER_COMMAND='if [[ "$DOCKER_REPLICA_SLOT" == *"1"* ]]; then export WORLD_NAME=$DR_MT_WORLD_NAME_1; export S3_YAML_NAME=$MT_S3_TRAINING_PARAMS_FILE_1; elif [[ "$DOCKER_REPLICA_SLOT" == *"2"* ]]; then export WORLD_NAME=$DR_MT_WORLD_NAME_2; export S3_YAML_NAME=$MT_S3_TRAINING_PARAMS_FILE_2; elif [[ "$DOCKER_REPLICA_SLOT" == *"3"* ]]; then export WORLD_NAME=$DR_MT_WORLD_NAME_3; export S3_YAML_NAME=$MT_S3_TRAINING_PARAMS_FILE_3; elif [[ "$DOCKER_REPLICA_SLOT" == *"4"* ]]; then export WORLD_NAME=$DR_MT_WORLD_NAME_4; export S3_YAML_NAME=$MT_S3_TRAINING_PARAMS_FILE_4; elif [[ "$DOCKER_REPLICA_SLOT" == *"5"* ]]; then export WORLD_NAME=$DR_MT_WORLD_NAME_5; export S3_YAML_NAME=$MT_S3_TRAINING_PARAMS_FILE_5; elif [[ "$DOCKER_REPLICA_SLOT" == *"6"* ]]; then export WORLD_NAME=$DR_MT_WORLD_NAME_6; export S3_YAML_NAME=$MT_S3_TRAINING_PARAMS_FILE_6; elif [[ "$DOCKER_REPLICA_SLOT" == *"7"* ]]; then export WORLD_NAME=$DR_MT_WORLD_NAME_7; export S3_YAML_NAME=$MT_S3_TRAINING_PARAMS_FILE_7; elif [[ "$DOCKER_REPLICA_SLOT" == *"8"* ]]; then export WORLD_NAME=$DR_MT_WORLD_NAME_8; export S3_YAML_NAME=$MT_S3_TRAINING_PARAMS_FILE_8; fi && ./run.sh multi distributed_training.launch'
```
Comment (larsll) [inline scripts/training/start.sh]: Let's have a look as to how we can combine this into something slightly more elegant. What we want to do is to pass in the WORLD_NAME and S3_YAML_NAME per worker, right? We can adapt the run.sh on the robomaker side to help on this too!

I would do the following; create a yaml or json and pass as separate (single) environment variable. We then adapt run.sh to push out its file based on the established WORKER_NUM which run.sh anyways needs to figure out.

**Review by larsll — CHANGES_REQUESTED** [review]

**Comment by larsll** [comment]
I created this one to match: https://github.com/aws-deepracer-community/deepracer-simapp/pull/51/files

File: scripts/training/prepare-config.py (hunk 1 above)
Comment (dartjason) [reply scripts/training/prepare-config.py]: Agreed, done

File: scripts/training/prepare-config.py (hunk 2 above)
Comment (dartjason) [reply scripts/training/prepare-config.py]: Moved file to tmp as requested, but have not been able to figure out right syntax to include DR_RUN_ID.

Comment (dartjason) [reply scripts/training/prepare-config.py]: this is now done

File: scripts/training/prepare-config.py [hunk 4]
```diff
@@ -86,3 +86,90 @@
     yaml.dump(config, yaml_file, default_flow_style=False, default_style='\'', explicit_start=True)
 
 s3_client.upload_file(Bucket=s3_bucket, Key=yaml_key, Filename=local_yaml_path)
```
Comment (larsll) [inline scripts/training/prepare-config.py]: Last comment - I think: The file without the number should only be uploaded if we are NOT having multi-config.

**Review by larsll — CHANGES_REQUESTED** [review]

File: scripts/training/prepare-config.py (hunk 4 above)
Comment (dartjason) [reply scripts/training/prepare-config.py]: yes ok.  once all is working properly, easiest probably just to move this down to an else statement after multi_config if statement.   need it here now to troubleshoot.

Comment (dartjason) [reply scripts/training/prepare-config.py]: this has now been completed

**Review by larsll — APPROVED** [review]

**Pull request merged**